api/src/opentrons/commands/publisher.py

```diff
@@ -1,154 +1,148 @@
+            message_creator_arg_names = set(message_creator_sig.parameters.keys())
+            message_creator_args = {
+                n: func_args[n] for n in message_creator_arg_names if n in func_args
+            }
+
+            # TODO (artyom, 20170927): we are doing this to be able to use
+            # the decorator in Instrument class methods, in which case
+            # self is effectively an instrument.
+            # To narrow the scope of this hack, we are checking if the
+            # command is expecting instrument first.
+            # We are also checking if call arguments have 'self' and
+            # don't have instruments specified, in which case
+            # instruments should take precedence.
+            if (
+                "instrument" in message_creator_arg_names
+                and "instrument" not in message_creator_args
+                and "self" in func_args
+            ):
+                message_creator_args["instrument"] = func_args["self"]
+
+            command_message = command(**message_creator_args)
+
+            with publish_context(broker=broker, command=command_message):
+                return func(*args, **kwargs)
 
+        return cast(FuncT, _decorated)
 
-FuncT = TypeVar("FuncT", bound=Callable[..., Any])
+    return _decorator
 
 
-def publish(command: CmdFunction, meta: Any = None) -> Callable[[FuncT], FuncT]:
-    """Publish events both before and after the decorated function has run."""
+@contextmanager
+def publish_context(broker: Broker, command: CommandPayload) -> Iterator[None]:
+    """Publish messages before and after the `with` block has run.
 
-    def _decorator(f: FuncT) -> FuncT:
-        @functools.wraps(
-            f, updated=functools.WRAPPER_UPDATES + ("__globals__",)  # type: ignore[operator]  # noqa: E501
-        )
-        def _decorated(*args: Any, **kwargs: Any) -> Any:
-            try:
-                broker = cast(Broker, args[0].broker)
-            except AttributeError:
-                raise RuntimeError(
-                    "Only methods of CommandPublisher \
-                    classes should be decorated."
-                )
-
-            do_publish(broker, command, f, "before", None, meta, *args, **kwargs)
-            res = f(*args, **kwargs)
-            do_publish(broker, command, f, "after", res, meta, *args, **kwargs)
-            return res
+    If an `error` is raised in the `with` block, it will be published in the "after"
+    message (if the ProtocolEngine is enabled) and re-raised.
+    """
+    capture_errors = feature_flags.enable_protocol_engine()
+    error = None
 
-        return cast(FuncT, _decorated)
+    _do_publish(broker=broker, command=command, when="before", error=None)
+    try:
+        yield
+    except Exception as e:
+        error = e
+        raise
+    finally:
+        if error is None or capture_errors is True:
+            _do_publish(broker=broker, command=command, when="after", error=error)
```

Comment:
This is a matter of taste, but I'd find this clearer if it were written like:

```suggestion
    try:
        yield
    except Exception as e:
        if capture_errors:
            _do_publish(broker=broker, command=command, when="after", error=e)
        raise
    else:
        _do_publish(broker=broker, command=command, when="after", error=None)
```

Because the syntax more naturally mirrors the decision tree that we're moving through here. The contained code either raised an exception or it didn't. If it did, then we publish with maybe that exception as an error, depending on the feature flag. If it didn't, we publish without an error.